### PR TITLE
[distributed] add gpu specs, % and absolute peak mem used per stage

### PR DESCRIPTION
**Pull request by lessw2020**
This PR:
1 - adds a GPUMemoryMonitor class to:

a - log the gpus' device capabilities at start of run (in yellow):
Screenshot 2024-09-08 at 5 23 59 PM

b - at the end of the run, show the absolute peak res memory as well as % peak memory used per stage:
Screenshot 2024-09-08 at 5 23 07 PM

### CODE REVIEW

File: dist_run.py [hunk 1]
```diff
@@ -231,6 +237,12 @@ def main():
     if pp_rank == pp_degree - 1 and tp_rank == 0:
         logger.info(f"Output: {output}")
 
+    # show peak memory stats for this stage
+    res_mem_gib, res_mem_pct = gpu_memory_monitor.get_peak_stats()
+    logger.info(
+        f"{color.blue} Memory used: {color.green}{res_mem_pct:.3f} %, {color.magenta}{res_mem_gib:.3f} GB{color.reset}"
+    )
+
```
Comment (kwen2501) [inline dist_run.py:245]: nit: maybe make the method "print_peak_stats()" if there is no other use of `res_mem_gib, res_mem_pct` than logging them?

File: distributed/utils.py [hunk 2]
```diff
@@ -76,3 +78,52 @@ class NoColor:
     cyan = ""
     white = ""
     reset = ""
+
+
+class GPUMemoryMonitor:
+    def __init__(self, device: str = "cuda:0"):
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/utils.py:84]: yes, fixed.

File: distributed/utils.py [hunk 3]
```diff
@@ -76,3 +78,52 @@ class NoColor:
...
+
+    def _to_pct(self, memory):
+        return 100 * memory / self.device_capacity
+
+    def get_peak_stats(self):
+        cuda_info = torch.cuda.memory_stats(self.device)
+
+        # max_active = cuda_info["active_bytes.all.peak"]
+        # max_active_gib = self._to_gib(max_active)
+        # max_active_pct = self._to_pct(max_active)
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/utils.py:110]: removed

File: distributed/utils.py [hunk 4]
```diff
@@ -76,3 +78,52 @@ class NoColor:
     cyan = ""
     white = ""
     reset = ""
+
+
+class GPUMemoryMonitor:
+    def __init__(self, device: str = "cuda:0"):
+        self.device = torch.device(device)  # device object
+        self.device_name = torch.cuda.get_device_name(self.device)
+        self.device_index = torch.cuda.current_device()
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/utils.py:87]: good catch.  I get the index now via self.device.index property to keep it consistent with passed in device.

File: distributed/utils.py [hunk 5]
```diff
@@ -76,3 +78,52 @@ class NoColor:
...
+
+
+def build_gpu_memory_monitor():
+    gpu_memory_monitor = GPUMemoryMonitor("cuda")
+    device_info = (
+        f"GPU capacity: {gpu_memory_monitor.device_name} ({gpu_memory_monitor.device_index}) "
+        f"with {gpu_memory_monitor.device_capacity_gib:.2f}GiB memory"
+    )
+
+    return gpu_memory_monitor, device_info
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply distributed/utils.py:129]: make sense.  I've refactored it to have the user invoke directly, and made the formatted info string returned via get_device_info().

File: dist_run.py (hunk 1 above)
Comment (lessw2020) [reply dist_run.py:245]: in this case, I wanted to return the data b/c if we want to setup tb logging or similar stress test monitoring, then we can pass this to summarywriter or other for dashboarding.

**Pull request merged**